### ISSUE TITLE
Rule proposal: `no-useless-array-method`

### ISSUE THREAD

**Issue by remcohaszing** [sindresorhus/eslint-plugin-unicorn#2045]
### Description

In various situations, array methods are best avoided, because they create an intermediate array for no benefit. Typically this is the use of `array.map()` or `array.filter()` followed by direct iteration or another array method.

### Fail

```js
array.map(fn1).map(fn2)
```

```js
array.filter(fn1).filter(fn2)
```

```js
array.map(fn1).forEach(fn2)
```

```js
array.filter(fn1).forEach(fn2)
```

```js
array.map(fn1).reduce(fn2)
```

```js
array.filter(fn1).reduce(fn2)
```

```js
array.filter(fn1).find(fn2)
```

```js
for (const item of array.map(fn1)) {
  fn2(item)
}
```

```js
for (const item of array.filter(fn1)) {
  fn2(item)
}
```

### Pass

```js
array.map(item => fn2(fn1(item)))
```

```js
array.filter(item => fn1(item) && fn2(item))
```

```js
array.forEach(item => {
  fn2(fn1(item))
})
```

```js
array.forEach((item) => {
  if (fn1(item)) {
    fn2(item)
  }
})
```

```js
array.reduce((acc, item) => {
  return fn2(acc, fn1(item))
})
```

```js
array.reduce((acc, item) => {
  if (fn1(item)) {
    return fn2(acc, item)
  }
  return acc
})
```

```js
array.find((item) => fn1(item) && fn2(item))
```

```js
for (const value of array) {
  const item = fn1(value)
  fn2(item)
}
```

```js
for (const item of array) {
  if (!fn1(item)) {
    continue
  }
  fn2(item)
}
```

### Additional Info

This could be autofixable.
